components/script/dom/gpudevice.rs

```diff
@@ -119,6 +158,53 @@ impl GPUDevice {
     pub fn id(&self) -> webgpu::WebGPUDevice {
         self.device
     }
+
+    pub fn handle_server_msg(&self, scope: ErrorScopeId, result: Result<(), GPUError>) {
+        let mut context = self.scope_context.borrow_mut();
+        let remove = if let Some(mut err_scope) = context.error_scopes.get_mut(&scope) {
+            err_scope.op_count -= 1;
+            match result {
+                Ok(()) => {},
+                Err(e) => {
+                    if err_scope.error.is_none() {
+                        err_scope.error = Some(e);
+                    }
+                },
+            }
+            if let Some(ref promise) = err_scope.promise {
+                if !promise.is_fulfilled() {
+                    if let Some(ref e) = err_scope.error {
+                        match e {
+                            GPUError::GPUValidationError(v) => promise.resolve_native(&v),
+                            GPUError::GPUOutOfMemoryError(w) => promise.resolve_native(&w),
+                        }
+                    } else if err_scope.op_count == 0 {
+                        promise.resolve_native(&None::<GPUError>);
+                    }
+                }
+            }
+            err_scope.op_count == 0 && err_scope.promise.is_some()
+        } else {
+            warn!("Could not find ErrroScope with Id({})", scope);
+            return;
+        };
+        if remove {
+            let _ = context.error_scopes.remove(&scope);
+        }
+    }
+
+    fn use_current_scope(&self) -> Option<ErrorScopeId> {
+        let mut context = self.scope_context.borrow_mut();
+        let scope_id = context.scope_stack.last().map(|s_id| *s_id);
```

Comment:
we shouldn't need that `map` at the end, since we are mapping right after that anyway

Reply:
I used `map` because contents of `context` had to be borrowed mutably in the next step. I couldn't do so if the previous borrow was alive. Therefore, I created a new `Option<u64>` via map here.

Reply:
ok, can also do `.cloned()`